Not building on Fedora 27

[mastag@FedoraWS build]$ cmake -DCMAKE_BUILD_TYPE=Debug ..
-- The C compiler identification is GNU 7.2.1
-- The CXX compiler identification is GNU 7.2.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- We are on a Linux system
-- Boost version: 1.64.0
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
--   system
--   filesystem
-- boost header: /usr/include
-- boost libs  : /usr/lib64/libboost_unit_test_framework.so;/usr/lib64/libboost_system.so;/usr/lib64/libboost_filesystem.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mastag/src/Zilliqa/build
[mastag@FedoraWS build]$ make
Scanning dependencies of target Common
[  1%] Building CXX object src/depends/common/CMakeFiles/Common.dir/Common.cpp.o
[  2%] Building CXX object src/depends/common/CMakeFiles/Common.dir/CommonData.cpp.o
[  2%] Building CXX object src/depends/common/CMakeFiles/Common.dir/CommonIO.cpp.o
[  3%] Building CXX object src/depends/common/CMakeFiles/Common.dir/FileSystem.cpp.o
[  4%] Building CXX object src/depends/common/CMakeFiles/Common.dir/FixedHash.cpp.o
[  4%] Building CXX object src/depends/common/CMakeFiles/Common.dir/RLP.cpp.o
[  5%] Building CXX object src/depends/common/CMakeFiles/Common.dir/SHA3.cpp.o
[  5%] Linking CXX shared library libCommon.so
[  5%] Built target Common
Scanning dependencies of target json_spirit
[  5%] Building CXX object src/depends/json_spirit/CMakeFiles/json_spirit.dir/json_spirit_reader.cpp.o
[  6%] Building CXX object src/depends/json_spirit/CMakeFiles/json_spirit.dir/json_spirit_value.cpp.o
[  7%] Building CXX object src/depends/json_spirit/CMakeFiles/json_spirit.dir/json_spirit_writer.cpp.o
[  7%] Linking CXX shared library libjson_spirit.so
[  7%] Built target json_spirit
Scanning dependencies of target Utils
[  7%] Building CXX object src/libUtils/CMakeFiles/Utils.dir/DataConversion.cpp.o
[  8%] Building CXX object src/libUtils/CMakeFiles/Utils.dir/Logger.cpp.o
[  9%] Building CXX object src/libUtils/CMakeFiles/Utils.dir/SanityChecks.cpp.o
[  9%] Building CXX object src/libUtils/CMakeFiles/Utils.dir/Scheduler.cpp.o
In file included from /home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:18:0:
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:37:26: fout: ‘std::function’ has not been declared
     void ScheduleAt(std::function<void (void)> f,
                          ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:37:34: fout: expected ‘,’ or ‘...’ before ‘<’ token
     void ScheduleAt(std::function<void (void)> f,
                                  ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:40:29: fout: ‘std::function’ has not been declared
     void ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds);
                             ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:40:37: fout: expected ‘,’ or ‘...’ before ‘<’ token
     void ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds);
                                     ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:42:36: fout: ‘std::function’ has not been declared
     void SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds);
                                    ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:42:44: fout: expected ‘,’ or ‘...’ before ‘<’ token
     void SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds);
                                            ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:76: fout: ‘function’ is not a member of ‘std’
     std::multimap<std::chrono::time_point<std::chrono::system_clock>, std::function<void (void)>> taskQueue;
                                                                            ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:76: note: suggested alternative: ‘is_function’
     std::multimap<std::chrono::time_point<std::chrono::system_clock>, std::function<void (void)>> taskQueue;
                                                                            ^~~~~~~~
                                                                            is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:76: fout: ‘function’ is not a member of ‘std’
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:76: note: suggested alternative: ‘is_function’
     std::multimap<std::chrono::time_point<std::chrono::system_clock>, std::function<void (void)>> taskQueue;
                                                                            ^~~~~~~~
                                                                            is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:96: fout: template argument 2 is invalid
     std::multimap<std::chrono::time_point<std::chrono::system_clock>, std::function<void (void)>> taskQueue;
                                                                                                ^~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.h:47:96: fout: template argument 4 is invalid
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp: In memberfunctie ‘void Scheduler::ServiceQueue()’:
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:38:30: fout: request for member ‘empty’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
             while (taskQueue.empty()) {
                              ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:42:31: fout: request for member ‘empty’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
             while (!taskQueue.empty()) {
                               ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:43:94: fout: request for member ‘begin’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
                 std::chrono::time_point<std::chrono::system_clock> timeToWaitFor = taskQueue.begin()->first;
                                                                                              ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:50:27: fout: request for member ‘empty’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
             if (taskQueue.empty())
                           ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:55:18: fout: ‘function’ is not a member of ‘std’
             std::function<void (void)> f = taskQueue.begin()->second;
                  ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:55:18: note: suggested alternative: ‘is_function’
             std::function<void (void)> f = taskQueue.begin()->second;
                  ^~~~~~~~
                  is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:55:27: fout: expected primary-expression before ‘void’
             std::function<void (void)> f = taskQueue.begin()->second;
                           ^~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:56:23: fout: request for member ‘erase’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
             taskQueue.erase(taskQueue.begin());
                       ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:56:39: fout: request for member ‘begin’ in ‘((Scheduler*)this)->Scheduler::taskQueue’, which is of non-class type ‘int’
             taskQueue.erase(taskQueue.begin());
                                       ^~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:60:17: fout: ‘f’ was not declared in this scope
                 f();
                 ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp: At global scope:
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:71:33: fout: variable or field ‘ScheduleAt’ declared void
 void Scheduler::ScheduleAt(std::function<void (void)> f, chrono::time_point<chrono::system_clock> t)
                                 ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:71:33: fout: ‘function’ is not a member of ‘std’
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:71:33: note: suggested alternative: ‘is_function’
 void Scheduler::ScheduleAt(std::function<void (void)> f, chrono::time_point<chrono::system_clock> t)
                                 ^~~~~~~~
                                 is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:71:42: fout: expected primary-expression before ‘void’
 void Scheduler::ScheduleAt(std::function<void (void)> f, chrono::time_point<chrono::system_clock> t)
                                          ^~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:71:99: fout: expected primary-expression before ‘t’
 void Scheduler::ScheduleAt(std::function<void (void)> f, chrono::time_point<chrono::system_clock> t)
                                                                                                   ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:80:36: fout: variable or field ‘ScheduleAfter’ declared void
 void Scheduler::ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                    ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:80:36: fout: ‘function’ is not a member of ‘std’
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:80:36: note: suggested alternative: ‘is_function’
 void Scheduler::ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                    ^~~~~~~~
                                    is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:80:45: fout: expected primary-expression before ‘void’
 void Scheduler::ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                             ^~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:80:69: fout: expected primary-expression before ‘deltaMilliSeconds’
 void Scheduler::ScheduleAfter(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                                                     ^~~~~~~~~~~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:85:59: fout: ‘std::function’ has not been declared
 static void SchedulePeriodicallyHelper(Scheduler* s, std::function<void (void)> f, int64_t deltaMilliSeconds)
                                                           ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:85:67: fout: expected ‘,’ or ‘...’ before ‘<’ token
 static void SchedulePeriodicallyHelper(Scheduler* s, std::function<void (void)> f, int64_t deltaMilliSeconds)
                                                                   ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp: In functie ‘void SchedulePeriodicallyHelper(Scheduler*, int)’:
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:87:5: fout: ‘f’ was not declared in this scope
     f();
     ^
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:88:62: fout: ‘deltaMilliSeconds’ was not declared in this scope
     s->ScheduleAfter(bind(&SchedulePeriodicallyHelper, s, f, deltaMilliSeconds), deltaMilliSeconds);
                                                              ^~~~~~~~~~~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:88:22: fout: ‘bind’ was not declared in this scope
     s->ScheduleAfter(bind(&SchedulePeriodicallyHelper, s, f, deltaMilliSeconds), deltaMilliSeconds);
                      ^~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:88:22: note: suggested alternative: ‘rand’
     s->ScheduleAfter(bind(&SchedulePeriodicallyHelper, s, f, deltaMilliSeconds), deltaMilliSeconds);
                      ^~~~
                      rand
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp: At global scope:
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:91:43: fout: variable or field ‘SchedulePeriodically’ declared void
 void Scheduler::SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                           ^~~~~~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:91:43: fout: ‘function’ is not a member of ‘std’
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:91:43: note: suggested alternative: ‘is_function’
 void Scheduler::SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                           ^~~~~~~~
                                           is_function
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:91:52: fout: expected primary-expression before ‘void’
 void Scheduler::SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                                    ^~~~
/home/mastag/src/Zilliqa/src/libUtils/Scheduler.cpp:91:76: fout: expected primary-expression before ‘deltaMilliSeconds’
 void Scheduler::SchedulePeriodically(std::function<void (void)> f, int64_t deltaMilliSeconds)
                                                                            ^~~~~~~~~~~~~~~~~
make[2]: *** [src/libUtils/CMakeFiles/Utils.dir/build.make:135: src/libUtils/CMakeFiles/Utils.dir/Scheduler.cpp.o] Fout 1
make[1]: *** [CMakeFiles/Makefile2:1422: src/libUtils/CMakeFiles/Utils.dir/all] Fout 2
make: *** [Makefile:84: all] Fout 2

--- a/src/libUtils/Scheduler.h
+++ b/src/libUtils/Scheduler.h
@@ -19,6 +19,7 @@
 
 #include <chrono>
 #include <condition_variable>
+#include <functional>
 #include <map>
 #include <mutex>
 #include <thread>

